api/v1alpha4/openstackcluster_webhook_test.go

```diff
@@ -111,6 +111,124 @@ func TestOpenStackCluster_ValidateUpdate(t *testing.T) {
 			},
 			wantErr: true,
 		},
+		{
+			name: "Toggle OpenStackCluster.Spec.Bastion.Enabled flag is allowed",
+			oldTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					CloudName: "foobar",
+					Bastion: &Bastion{
+						Instance: OpenStackMachineSpec{
+							CloudName: "foobar",
+							Image:     "foobar",
+							Flavor:    "minimal",
+						},
+						Enabled: false,
+					},
+				},
+			},
+			newTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					CloudName: "foobar",
+					Bastion: &Bastion{
+						Instance: OpenStackMachineSpec{
+							CloudName: "foobarbaz",
+							Image:     "foobarbaz",
+							Flavor:    "medium",
+						},
+						Enabled: true,
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "Changing empty OpenStackCluster.Spec.Bastion is allowed",
+			oldTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					CloudName: "foobar",
+				},
+			},
+			newTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					CloudName: "foobar",
+					Bastion: &Bastion{
+						Instance: OpenStackMachineSpec{
+							CloudName: "foobar",
+							Image:     "foobar",
+							Flavor:    "medium",
+						},
+						Enabled: true,
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "Changing OpenStackCluster.Spec.Bastion with no deployed bastion host is allowed",
+			oldTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					CloudName: "foobar",
+					Bastion: &Bastion{
+						Instance: OpenStackMachineSpec{
+							CloudName: "foobar",
+							Image:     "foobar",
+							Flavor:    "minimal",
+						},
+						Enabled: false,
+					},
+				},
+				Status: OpenStackClusterStatus{},
+			},
+			newTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					CloudName: "foobar",
+					Bastion: &Bastion{
+						Instance: OpenStackMachineSpec{
+							CloudName: "foobarbaz",
+							Image:     "foobarbaz",
+							Flavor:    "medium",
+						},
+						Enabled: true,
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "Changing OpenStackCluster.Spec.Bastion with deployed bastion host is not allowed",
+			oldTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					CloudName: "foobar",
+					Bastion: &Bastion{
+						Instance: OpenStackMachineSpec{
+							CloudName: "foobar",
+							Image:     "foobar",
+							Flavor:    "minimal",
+						},
+						Enabled: true,
+					},
+				},
+				Status: OpenStackClusterStatus{
+					Bastion: &Instance{
+						Name: "foobar",
+					},
+				},
+			},
+			newTemplate: &OpenStackCluster{
+				Spec: OpenStackClusterSpec{
+					CloudName: "foobar",
+					Bastion: &Bastion{
+						Instance: OpenStackMachineSpec{
+							CloudName: "foobarbaz",
+							Image:     "foobarbaz",
+							Flavor:    "medium",
+						},
+						Enabled: true,
+					},
+				},
+			},
+			wantErr: true,
+		},
```

Comment:
I think you're missing an interesting test here: flipping Enabled when the Bastion exists in status is allowed.

Reply:
Good point. I added also another test `Disabling and changing the OpenStackCluster.Spec.Bastion while it's running is allowed`. This should be fine, as the bastion host will be destroyed anyway.